lua/telescope/previewers.lua

```diff
@@ -443,41 +561,90 @@ previewers.vimgrep = defaulter(function(opts)
   }
 end, {})
 
-previewers.ctags = defaulter(function(_)
+previewers.internal_cat = defaulter(function(_)
+  return previewers.new_buffer_previewer {
+    get_buffer_by_name = function(self, entry)
+      return from_entry.path(entry, true)
+    end,
+
+    define_preview = function(self, entry, status)
+      with_preview_window(status, nil, function()
+        local path = from_entry.path(entry, true)
+        if path == nil or path == '' then return end
+        file_maker_async(path, self.state.bufnr, self.state.bufname)
+      end)
+    end
+  }
+end, {})
+
+previewers.internal_vimgrep = defaulter(function(_)
   return previewers.new_buffer_previewer {
     setup = function()
-      return {}
+      return {
+        last_set_bufnr = nil
+      }
     end,
 
+    teardown = function(self)
+      if self.state and self.state.hl_id then
+        vim.api.nvim_buf_clear_namespace(self.state.last_set_bufnr, previewer_ns, 0, -1)
+      end
+    end,
+
+    get_buffer_by_name = function(self, entry)
+      return from_entry.path(entry, true)
+    end,
+
+    define_preview = function(self, entry, status)
+      with_preview_window(status, nil, function()
+        local lnum = entry.lnum or 0
+        local path = from_entry.path(entry, true)
+        if path == nil or path == '' then return end
+
+        file_maker_sync(path, self.state.bufnr, self.state.bufname)
+
+        if lnum ~= 0 then
+          if self.state.last_set_bufnr then
+            pcall(vim.api.nvim_buf_clear_namespace, self.state.last_set_bufnr, previewer_ns, 0, -1)
+          end
+          pcall(vim.api.nvim_buf_add_highlight, self.state.bufnr, previewer_ns, "Search", lnum - 1, 0, -1)
+          pcall(vim.api.nvim_win_set_cursor, status.preview_win, {lnum, 0})
+        end
+
+        self.state.last_set_bufnr = self.state.bufnr
+      end)
+    end
+  }
+end, {})
+
+previewers.internal_qflist = previewers.internal_vimgrep
```

Comment:
For naming sakes :)

Reply:
:+1: Also when there are a need to adjust the previewer for qflist, we wouldn't introduce a new api